src/cascadia/TerminalApp/PowershellCoreProfileGenerator.cpp

```diff
@@ -10,81 +10,320 @@
+// Return value:
+// - a comprehensive priority-ordered list of powershell instances.
+static std::vector<PowerShellInstance> _collectPowerShellInstances()
+{
+    std::vector<PowerShellInstance> versions;
+
+    _accumulateTraditionalLayoutPowerShellInstancesInDirectory(L"%ProgramFiles%\\PowerShell", PowerShellFlags::None, versions);
+
+#if defined(_M_AMD64) || defined(_M_ARM64) // No point in looking for WOW if we're not somewhere it exists
+    _accumulateTraditionalLayoutPowerShellInstancesInDirectory(L"%ProgramFiles(x86)%\\PowerShell", PowerShellFlags::WOWx86, versions);
+#endif
+
+#if defined(_M_ARM64) // no point in looking for WOA if we're not on ARM64
+    _accumulateTraditionalLayoutPowerShellInstancesInDirectory(L"%ProgramFiles(Arm)%\\PowerShell", PowerShellFlags::WOWARM, versions);
+#endif
+
+    _accumulateStorePowerShellInstances(versions);
+
+    _accumulatePwshExeInDirectory(L"%USERPROFILE%\\.dotnet\\tools", PowerShellFlags::Dotnet, versions);
+    _accumulatePwshExeInDirectory(L"%USERPROFILE%\\scoop\\shims", PowerShellFlags::Scoop, versions);
+
+    std::sort(versions.rbegin(), versions.rend()); // sort in reverse (best first)
+
+    return versions;
+}
+
+// Legacy GUIDs:
+//   - PowerShell Core       574e775e-4f2a-5b96-ac1e-a2962a402336
+static constexpr GUID PowershellCoreGuid{ 0x574e775e, 0x4f2a, 0x5b96, { 0xac, 0x1e, 0xa2, 0x96, 0x2a, 0x40, 0x23, 0x36 } };
+
+std::wstring_view PowershellCoreProfileGenerator::GetNamespace()
+{
+    return PowershellCoreGeneratorNamespace;
+}
+
+// Method Description:
+// - Checks if pwsh is installed, and if it is, creates a profile to launch it.
+// Arguments:
+// - <none>
+// Return Value:
+// - a vector with the PowerShell Core profile, if available.
+std::vector<TerminalApp::Profile> PowershellCoreProfileGenerator::GenerateProfiles()
+{
+    std::vector<TerminalApp::Profile> profiles;
+
+    auto psInstances = _collectPowerShellInstances();
+    for (const auto& psI : psInstances)
+    {
+        auto name = psI.Name();
+        auto profile{ CreateDefaultProfile(name) };
+        profile.SetCommandline(psI.executablePath.wstring());
+        profile.SetStartingDirectory(DEFAULT_STARTING_DIRECTORY);
+        profile.SetColorScheme({ L"Campbell" });
+
+        profile.SetIconPath(WI_IsFlagSet(psI.flags, PowerShellFlags::Preview) ? POWERSHELL_PREVIEW_ICON : POWERSHELL_ICON);
+        profiles.emplace_back(std::move(profile));
+    }
+
+    if (profiles.size() > 0)
+    {
+        // Give the first ("algorithmically best") profile the official, and original, "PowerShell Core" GUID.
+        // This will turn the anchored default profile into "PowerShell Core Latest for Native Architecture through Store"
+        // (or the closest approximation thereof). It may choose a preview instance as the "best" if it is a higher version.
+        auto firstProfile = profiles.begin();
+        firstProfile->SetGuid(PowershellCoreGuid);
+        firstProfile->SetName(L"PowerShell");

```

Comment:
Do we really want to rename it? When this is generated for me, how will I know what version it actually is?

Reply:
That's a good question that I don't have a good answer for. I wanted to rename it because _when we put the name in the profile, it will be permanently branded_. If we later detect a "better" version (like 7 goes out of preview and becomes GA), it would still say "PowerShell 7 Preview" but when they launch it it'll be PowerShell 7 GA.

If we don't replace the base powershell core GUID, their default profile will just disappear one day.

If we added a profile _comment_, we could omit the names from the json for certain dynamic profiles. The comment would contain more information, like the version and where it was detected... but _that would also be permanently fixed in time_ 😦

Reply:
okay good enough for me